client/src/main/java/io/pravega/client/byteStream/ByteStreamReader.java

```diff
@@ -0,0 +1,119 @@
+/**
+ * Copyright (c) 2018 Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.client.byteStream;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.ByteBuffer;
+import java.nio.channels.AsynchronousChannel;
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * Allows for reading raw bytes from a segment. This class is designed such that it can be used with
+ * or without blocking. To avoid blocking use the {@link #onDataAvailable()} method to make sure to
+ * only call {@link #read(byte[])} when there is data {@link #available()}.
+ *
+ * It is safe to invoke methods on this class from multiple threads, but doing so will not increase
+ * performance.
+ */
+public abstract class ByteStreamReader extends InputStream implements AsynchronousChannel, AutoCloseable {
+
+    /**
+     * Returns the current byte offset in the segment. This call does not block.
+     */
+    public abstract long getOffset();
+
+    /**
+     * Jumps to the provided offset. Future read calls will read from this position. This makes a
+     * synchronous RPC to the server to validate the offset provided.
+     * 
+     * @param offset The offset to jump to.
+     * @throws InvalidOffsetException If the offset provided does not exist in the segment.
+     */
+    public abstract void jumpToOffset(long offset) throws InvalidOffsetException;
+
+    /**
+     * Returns the number of bytes that can be read without blocking. If the number returned is > 0
+     * then a call to {@link #read(byte[]))} will return data from memory without blocking. If the
+     * number returned is 0 then {@link #read(byte[]))} will block If -1 is returned this indicates
+     * the end of the stream has been reached and a call to {@link #read(byte[])} will return -1.
+     * 
+     * @see java.io.InputStream#available()
+     */
+    @Override
+    public abstract int available();
+
+    /**
+     * This make a RPC to the server to fetch the highest offset in the segment.
+     */
+    public abstract long fetchTailOffset();
+
+    /**
+     * Don't call this. It is very wasteful.
+     */
+    @Override
+    public abstract int read() throws IOException;
+
+    /**
+     * See {@link InputStream#read(byte[])}. This is equivlent to calling
+     * {@code read(b, 0, b.length) }
+     * 
+     * Will only block if {@link #available()} is 0.
+     */
+    @Override
+    public abstract int read(byte[] b) throws IOException;
```

Comment:
Do we have a limit on the length of the read or can it be arbitrarily large? We should document what it is in the javadoc.

Reply:
You can give us as large of a buffer as you want to fill, but it is not guaranteed that we will fill it. We will only ever return what is in memory which maxes out at 1.25mb but in practice there isn't a performance advantage to reading in chunks larger than .25 mb, but these are internal constants that may change. I don't thing we need to explicitly document them as they have no direct impact on the user, if they want say 8mb at a time they can just call this method in a loop and eventually their buffer will be filled.

Reply:
Don't you think that the application developer will need to know that to send 8mb, it needs to be done in a loop rather than sending it at once? We need to explain how a developer is supposed to use the API. If we change it later, then we update the javadocs/docs.

Reply:
This pattern is standard in `InputStream`. This method is overridden from a base class. Let's not overload the documentation too.

Reply:
I'm actually ok with it, the application should not expect to read all at once in any case, and it doesn't induce an error to read with a larger buffer.